### DISCUSSION TITLE
How to setup a generic object mapping from entities to EntityRepository inference

### DISCUSSION THREAD

**Discussion by ryankauk** [mikro-orm/mikro-orm#1683] [Q&A] [answered]
I'm wondering if I could get a bit of guidance on using generics mapping from entities to repository.

Here's what I have.

```typescript
export const createDB2 = <
  Entities extends { [k in Key]: AnyEntity<Entities[Key]> },
  Key extends string = string
>(
  entity: { [k in Key]: EntityName<Entities[Key]> }
) => {
  const orm: MikroORM<MongoDriver> = {} as any;
  return {} as {
    [key in Key]: MongoEntityRepository<Entities[key]>;
  };
};
const repos = createDB2({ MyEntity: MyEntity });
repos.MyEntity.findOne({});
// (property) MyEntity: MongoEntityRepository<AnyEntity<unknown>>
```

however it works properly if i directly declare the types instead of inferring

```typescript
const repos = createDB2<{ MyEntity: MyEntity }, "MyEntity">({ MyEntity: MyEntity });
repos.MyEntity.findOne({});
// (property) MyEntity: MongoEntityRepository<MyEntity>
```

**Comment by B4nan** [comment, marked as answer]
Well but in your first example, you don't have the `Key` information anywhere, it will always fallback to the default value, so you end up with `unknown` as a result. There are more issues in your code, like `Entities[Key]` should be `Entities[k]` I guess? Also I don't see a reason to use `AnyEntity` here. 

The code is quite messy, so instead of debugging it more, here is a different approach:

```ts
type Ret<T> = { [k in keyof T]: T[k] extends EntityName<infer U> ? EntityRepository<U> : EntityRepository<unknown> };

function createDB<T>(map: T): Ret<T> {
  return map as any;
}

const repos = createDB({ MyEntity: MyEntity });
repos.MyEntity.findOne();
// (property) MyEntity: EntityRepository<MyEntity>
```

https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=15&pc=51#code/C4TwDgpgBAogdsAlqAwgGwIYGcsB4AqAfFALxQBiArnAMZID2cUAZFAN5RgBO9wv4EAFxR8UAL4BuAFChIsBMhAA5DAFsIBYmSzAuiOAHMoAH3lJUmHJumzo8cyABKEMPSzJ6XEJtLsoAM30AEwB5OAgACgBKYVFJKSkaSywoAFkQe0U-f3p6AH5hHT1DCXEE2yhnYB8yDgBtAGsofSgGiBB6fxEAXVjG7qgIAA9gCDgglMzQFXVcfX8ILigAVWI8s0VnV3c+L1xVqGEppxc3Dz3qBrh6AHc4Ynipf2o6REYoGi4IDFGAEQAhTQRVQYMCxGKVCDVIjsKRQKBfYCULhMEFgKDYDFwEDSMQJGiMHQI04pMifb5-f4RDjpY7CWkKUDiKLSL7bAB0DIc7MC4zCkRZUgA9EKoBFuPRIFxQFE0hlGSAjgqtmddt4uYpCEA

**Reply by ryankauk** [reply]
Thank you so much! sorry for the mess I'm not that good with generics yet.
